cheatsheets/Protect_FileUpload_Against_Malicious_File.md

```diff
@@ -1,597 +1,138 @@
-                            }
-                            case "BMP": {
-                                imageParser = new BmpImageParser();
-                                break;
-                            }
-                            case "GIF": {
-                                imageParser = new GifImageParser();
-                                break;
-                            }
-                            case "PNG": {
-                                imageParser = new PngImageParser();
-                                break;
-                            }
-                            case "WBMP": {
-                                imageParser = new WbmpImageParser();
-                                break;
-                            }
-                            case "XBM": {
-                                imageParser = new XbmImageParser();
-                                break;
-                            }
-                            case "XPM": {
-                                imageParser = new XpmImageParser();
-                                break;
-                            }
-                            default: {
-                                throw new IOException("Format of the original image is not" + 
-                                                      " supported for write operation !");
-                            }
-
-                        }
-                        imageParser.writeImage(sanitizedImage, fos, new HashMap<>());
-                    }
-
-                }
-
-                // Set state flag
-                safeState = true;
-            }
-        } catch (Exception e) {
-            safeState = false;
-            LOG.warn("Error during Image file processing !", e);
-        }
-
-        return safeState;
-    }
-}
+```python
+# This code snippet only serves as an example, and should not be used
+blacklist = ["php", "php5"]  # can be replaced with a regex pattern, but that would be even weaker, such as \.php$
+for bad in blacklist:
+    if "file.php".endswith(bad):
+        // raise exception
 ```
 
-# Sources of the prototype
+### User Permissions
+
+Before any file upload service is accessed, proper validation should occur on two levels for the user uploading a file:
+
+- Authentication level
+  - The user should be a registered user, or an identifiable user, in order to set restrictions and limitations for their upload capabilities
+- Authorization level
+  - Make sure that the user's accesses allow them to upload a file, or multiple files for administrative cases.
+
+### Content-Type Validation
+
+Other than defining the extension of the uploaded file, its MIME-type can be checked for a quick protection against simple file upload attacks.
+
+This can be done preferrably in a whitelist approach; otherwise, this can be done in a blacklist approach.
+
+> This should not be used on its own, as bypassing it is pretty common and easy.
+
+### File Name Sanitization
```

Comment:
I think this section needs to be expanded a bit. There a few different issues here that should be split into (sub)sections:

- Validating the file name that the user provides when uploading.
  - Length
  - Dangerous characters (null bytes, directory traversal, colons, slashes, etc)
  - File extension (subsection?)
- What name (if any) the file should be stored under.
  - Never use the provided filename unless you have to
  - Avoid predictable names for user-facing files (incremented numbers, timestamps, hashes, etc)
  - Dangerous characters
  - Dangerous filenames (`CON`, `AUX`, `.htaccess`, etc)
- What name the file should be served under
  - Dangerous characters like CRLF (response splitting/header injection)
  - Dangerous extensions
  - Content-type disposition
  - Browsers doing stupid things like MIME sniffing

Reply:
MIME sniffing occurs on the loading of the file client-side, and not on the upload. I added MSDN and Wikipedia's information to this. I doubt I can cover what you wrote in a CS without repeating what is already mentioned in there.